### ISSUE TITLE
Implement training metrics logging

### ISSUE THREAD

**Issue by tgrunnagle** [tgrunnagle/go-towerfall#45]
## Overview

Implement a training metrics logging system for the ML bot training pipeline. This component will log training metrics including rewards, episode lengths, and win rates to files and/or TensorBoard for monitoring and analysis.

Part of: #14 (ML Bot Training Epic)  
Parent Task: TASK-P008 (Spectator & Monitoring)  
Task ID: TASK-031

## Dependencies

- TASK-022 (#40) - PPO training loop (provides the metrics to log)

## Background

During RL training, it's essential to track key metrics to understand training progress, diagnose issues, and compare model generations. The training metrics logging system should integrate seamlessly with the PPO training loop implemented in TASK-022 (#40) and provide both file-based and TensorBoard logging options.

From the requirements:
- The successive training system needs to track metrics to determine when a model is "better" than its predecessor (average kills/deaths > previous generation)
- Spectator and monitoring features are needed to observe training progress

### Functional Requirements

1. **Core Metrics to Log**
   - Episode rewards (total reward per episode)
   - Episode lengths (timesteps per episode)
   - Win rates (calculated over rolling window)
   - Kill/death ratios
   - Average kills per episode
   - Average deaths per episode

2. **Logging Destinations**
   - File-based logging (JSON/CSV format for easy analysis)
   - TensorBoard integration for real-time visualization
   - Both should be configurable/optional

3. **Logging Frequency**
   - Per-episode summary metrics
   - Periodic aggregate statistics (configurable interval)
   - Training step-level metrics (loss values, learning rate, etc.)

4. **Metrics Categories**
   - **Training metrics**: policy loss, value loss, entropy, KL divergence
   - **Performance metrics**: rewards, episode lengths, win rates
   - **Game metrics**: kills, deaths, K/D ratio, accuracy

### Non-Functional Requirements

1. **Performance**: Logging should not significantly impact training speed
2. **Configurability**: Easy to enable/disable specific metrics or destinations
3. **Compatibility**: Should work with the hyperparameter configuration system (TASK-024, #28)

### File Structure

```
bot2/
├── training/
│   ├── __init__.py
│   ├── metrics/
│   │   ├── __init__.py
│   │   ├── logger.py          # Main MetricsLogger class
│   │   ├── writers/
│   │   │   ├── __init__.py
│   │   │   ├── base.py        # Abstract writer interface
│   │   │   ├── file_writer.py # JSON/CSV file writer
│   │   │   └── tensorboard_writer.py  # TensorBoard writer
│   │   └── aggregators.py     # Rolling statistics computation
```

### Core Components

1. **MetricsLogger Class**
   ```python
   class MetricsLogger:
       def __init__(self, log_dir: str, writers: list[MetricsWriter] | None = None):
           ...
       
       def log_episode(self, episode_data: EpisodeMetrics) -> None:
           """Log metrics for a completed episode."""
           ...
       
       def log_training_step(self, step_data: TrainingStepMetrics) -> None:
           """Log metrics for a training step (PPO update)."""
           ...
       
       def log_scalar(self, tag: str, value: float, step: int) -> None:
           """Log a single scalar metric."""
           ...
       
       def get_summary(self) -> dict[str, float]:
           """Get aggregated summary statistics."""
           ...
       
       def close(self) -> None:
           """Flush and close all writers."""
           ...
   ```

2. **Metrics Data Classes (Pydantic Models)**
   ```python
   class EpisodeMetrics(BaseModel):
       episode_id: int
       total_reward: float
       length: int
       kills: int
       deaths: int
       win: bool
       timestamp: datetime
   
   class TrainingStepMetrics(BaseModel):
       step: int
       policy_loss: float
       value_loss: float
       entropy: float
       kl_divergence: float
       learning_rate: float
   
   class AggregateMetrics(BaseModel):
       mean_reward: float
       mean_length: float
       win_rate: float
       mean_kd_ratio: float
       episodes_count: int
   ```

3. **MetricsWriter Interface**
   ```python
   class MetricsWriter(ABC):
       @abstractmethod
       def write_scalar(self, tag: str, value: float, step: int) -> None: ...
       
       @abstractmethod
       def write_scalars(self, main_tag: str, tag_scalar_dict: dict[str, float], step: int) -> None: ...
       
       @abstractmethod
       def flush(self) -> None: ...
       
       @abstractmethod
       def close(self) -> None: ...
   ```

### Integration Points

1. **With PPO Training Loop (TASK-022)**
   - The PPO training loop should call `log_episode()` after each episode
   - The PPO training loop should call `log_training_step()` after each policy update
   - Metrics logger instance should be passed to or accessible from the training orchestrator

2. **With Hyperparameter Config (TASK-024)**
   - Logging configuration should be part of the training config
   - Options for log directory, writer selection, logging frequency

3. **With Model Registry (TASK-026, future)**
   - Aggregate metrics should be stored alongside model checkpoints
   - Used for determining model "better" threshold in successive training

### TensorBoard Integration

```python
from torch.utils.tensorboard import SummaryWriter

class TensorBoardWriter(MetricsWriter):
    def __init__(self, log_dir: str):
        self.writer = SummaryWriter(log_dir)
    
    def write_scalar(self, tag: str, value: float, step: int) -> None:
        self.writer.add_scalar(tag, value, step)
    
    # ... other methods
```

### File-Based Logging

```python
class FileWriter(MetricsWriter):
    def __init__(self, log_dir: str, format: str = "json"):
        self.log_dir = Path(log_dir)
        self.format = format
        self.buffer: list[dict] = []
    
    def write_scalar(self, tag: str, value: float, step: int) -> None:
        self.buffer.append({"tag": tag, "value": value, "step": step})
    
    def flush(self) -> None:
        # Write buffer to file
        ...
```

## Acceptance Criteria

1. [ ] `MetricsLogger` class implemented with support for multiple writers
2. [ ] `TensorBoardWriter` implemented and functional
3. [ ] `FileWriter` implemented with JSON and CSV format support
4. [ ] Pydantic models defined for `EpisodeMetrics`, `TrainingStepMetrics`, `AggregateMetrics`
5. [ ] Rolling statistics aggregation for win rate, mean reward, etc.
6. [ ] Integration tested with the PPO training loop from TASK-022
7. [ ] Unit tests for all logging components
8. [ ] Configuration options documented and integrated with training config
9. [ ] Log files are properly rotated/managed to prevent disk space issues
10. [ ] TensorBoard logs can be visualized with `tensorboard --logdir <path>`

## Testing Strategy

1. **Unit Tests**
   - Test each writer independently
   - Test aggregation calculations
   - Test Pydantic model validation

2. **Integration Tests**
   - Test MetricsLogger with multiple writers
   - Test integration with mock training loop
   - Verify TensorBoard output format

3. **Manual Verification**
   - Run training with logging enabled
   - Verify TensorBoard visualization works
   - Verify file outputs are correct format

## Notes

- Use `torch.utils.tensorboard.SummaryWriter` for TensorBoard integration
- Consider adding async writing for file logger to minimize I/O blocking
- Rolling window size for aggregate metrics should be configurable (default: 100 episodes)
- Follow project standards: use native Python typing, relative imports from bot2 directory
